src/macros-by-example.md

```diff
@@ -24,149 +24,428 @@
+path), it's first looked up in textual scoping. If this does not yield any
+results, then it is looked up in path-based scoping. If the macro's name is
+qualified with a path, then it is only looked up in path-based scoping.
+
+```rust,ignore
+use lazy_static::lazy_static; // Path-based import.
+
+macro_rules! lazy_static { // Textual definition.
+    (lazy) => {};
+}
+
+lazy_static!{lazy} // Textual lookup finds our macro first.
+self::lazy_static!{} // Path-based lookup ignores our macro, finds imported one.
+```
+
+### Textual Scope
+
+Textual scope is based largely on the order that things appear in source files,
+and works similarly to the scope of local variables declared with `let` except
+it also applies at the module level. When `macro_rules!` is used to define a
+macro, the macro enters scope after the definition (note that it can still be
+used recursively, since names are looked up from the invocation site), up until
+its surrounding scope, typically a module, is closed. This can enter child
+modules and even span across multiple files:
+
+```rust,ignore
+//// src/lib.rs
+mod has_macro {
+    // m!{} // Error: m is not in scope.
+
+    macro_rules! m {
+        () => {};
+    }
+    m!{} // OK: appears after declaration of m.
+
+    mod uses_macro;
+}
+
+// m!{} // Error: m is not in scope.
+
+//// src/has_macro/uses_macro.rs
+
+m!{} // OK: appears after delcaration of m in src/lib.rs
+```
+
+It is not an error to define a macro multiple times; the most recent declaration
+will shadow the previous one unless it has gone out of scope.
+
+```rust
+macro_rules! m {
+    (1) => {};
+}
+
+m!(1);
+
+mod inner {
+    m!(1);
+
+    macro_rules! m {
+        (2) => {};
+    }
+    // m!(1); // Error: no rule matches '1'
+    m!(2);
+
+    macro_rules! m {
+        (3) => {};
+    }
+    m!(3);
+}
+
+m!(1);
+```
+
+Macros can be declared and used locally inside functions as well, and work
+similarly:
+
+```rust
+fn foo() {
+    // m!(); // Error: m is not in scope.
+    macro_rules! m {
+        () => {};
+    }
+    m!();
+}
+
+
+// m!(); // Error: m is not in scope.
+```
+
+The `#[macro_use]` attribute has two purposes. First, it can be used to make a
+module's macro scope not end when the module is closed, by applying it to a
+module:
+
+```rust
+#[macro_use]
+mod inner {
+    macro_rules! m {
+        () => {};
+    }
+}
+
+m!();
+```
+
+Second, it can be used to import macros from another crate, by attaching it to
+an  `extern crate` declaration appearing in the crate's root module. Macros
+imported this way are imported into the prelude of the crate, not textually,
+which means that they can be uare shadowed by any other name. While macros
+imported by `#[macro_use]` can be used before the import statement, in case of a
+conflict, the last macro imported wins. Optionally, a list of macros to import
+can be specified; this is not supported when `#[macro_use]` is applied to a
```

Comment:
We're documenting the rules as-is here so this is irrelevant for the reference, but is there a reason for this restriction? cc @petrochenkov

Reply:
I wish `#[macro_use]` on `mod` was still named `#[macro_escape]` to avoid unnecessary analogies.
`#[macro_escape(name)]` would be a new feature rather than lifted restriction.
I suspect the most common answer to the question "why is feature X not in the language" applies here as well - nobody proposed and implemented it.

Reply:
Ah; I thought there was some technical reason rather than a slip... since there doesn't seem to be, I think it would be good to simply fix this as an "omission" so that we can remove this line from the reference and simplify the language...